src/main/scala/cognite/spark/v1/FlexibleDataModelCorePropertySyncRelation.scala

```diff
@@ -110,54 +110,111 @@ private[spark] class FlexibleDataModelCorePropertySyncRelation(
+    * If the cursor has expired, generate a new cursor and do a full back fill.
+    */
+  private def generateFutureItemsCursor(
+      cursors: Option[Map[String, String]],
+      instanceType: InstanceType,
+      select: Map[String, SelectExpression]): IO[(Option[String], Boolean)] =
+    fetchData(
+      isBackFill = false,
+      cursors = cursors,
+      `with` = Map(
+        "sync" -> generateTableExpression(instanceType, FilterDefinition.Not(MatchAll(JsonObject())))),
+      select = select)
+      .map { sr =>
+        (sr.nextCursor, cursors.isEmpty)
+      }
+      .redeemWith(
+        {
+          case e: CdpApiException if e.code == 400 && e.message.startsWith("Cursor has expired") =>
+            logger.warn("Cursor has expired, doing a full backfill")
+            generateFutureItemsCursor(None, instanceType, select)
+          case e => IO.raiseError(e)
+        },
+        IO.pure
+      )
+
+  private def generateTableExpression(
+      instanceType: InstanceType,
+      filters: FilterDefinition): TableExpression =
+    instanceType match {
+      case InstanceType.Edge =>
+        TableExpression(edges = Some(EdgeTableExpression(filter = Some(filters))))
+      case InstanceType.Node =>
+        TableExpression(nodes = Some(NodesTableExpression(filter = Some(filters))))
+    }
 
-    Seq(syncOut(instanceSyncRequest, selectedColumns))
+  private def fetchData(
+      isBackFill: Boolean,
+      cursors: Option[Map[String, String]],
+      `with`: Map[String, TableExpression],
+      select: Map[String, SelectExpression]): IO[ItemsWithCursor[InstanceDefinition]] = {
+    val response = if (isBackFill) {
+      client.instances.queryRequest(
+        InstanceQueryRequest(
+          `with` = `with`,
+          cursors = cursors,
+          select = select,
+          includeTyping = Some(true)
+        )
+      )
+    } else {
+      client.instances.syncRequest(
+        InstanceSyncRequest(
+          `with` = `with`,
+          cursors = cursors,
+          select = select,
+          includeTyping = Some(true)
+        )
+      )
+    }
+    response.map { qr =>
+      val itemDefinitions = qr.items.getOrElse(Map.empty).get("sync").getOrElse(Vector.empty)
+      val nextCursor = qr.nextCursor.getOrElse(Map.empty).get("sync")
+      ItemsWithCursor(itemDefinitions, nextCursor)
+    }
   }
   // scalastyle:on cyclomatic.complexity
 
-  private def syncOut(syncRequest: InstanceSyncRequest, selectedProps: Array[String])(
+  private def syncOut(
+      futureItems: Option[String],
+      cursors: Option[Map[String, String]],
+      `with`: Map[String, TableExpression],
+      select: Map[String, SelectExpression],
+      selectedProps: Array[String])(
       implicit F: Async[IO]): Stream[IO, ProjectedFlexibleDataModelInstance] =
     Stream.eval {
-      syncWithCursor(syncRequest).map { sr =>
+      fetchData(futureItems.nonEmpty, cursors, `with`, select).map { sr =>
```

Comment:
How is presence of future items cursor related to backfill? generateFutureItemsCursor seems to base _2 returned boolean independently of returning the cursor?

It feels like some of confusion could be avoided if different wrapper types were used for /sync and /query cursors and if pairs of options etc were converted to case classes that represent different modes..

Reply:
yeah, I think there was a bug, did some refactoring